## Abstract

The data is related to **direct marketing campaigns** (phone calls) of a Portuguese banking institution.  
The classification goal is to **predict if a client will subscribe to a term deposit** (`y`).

**A term deposit is a type of savings account where you deposit money for a fixed period of time—called the term—and earn a guaranteed interest rate. You agree not to withdraw the money until the term ends, also known as the maturity date.** (`y`).

---

## Data Set Information

The marketing campaigns were based on phone calls. Often, more than one contact was required to determine whether the product (bank term deposit) would be subscribed (`'yes'` or `'no'`).

---

## Attribute Information

### 📂 Bank Client Data
- **age** *(numeric)*: Client's age  
- **job** *(categorical)*: Job type  
  `['admin.', 'blue-collar', 'entrepreneur', 'housemaid', 'management', 'retired', 'self-employed', 'services', 'student', 'technician', 'unemployed', 'unknown']`
- **marital** *(categorical)*: Marital status  
  `['divorced', 'married', 'single', 'unknown']`
- **education** *(categorical)*: Education level  
  `['basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'illiterate', 'professional.course', 'university.degree', 'unknown']`
- **default** *(categorical)*: Has credit in default?  
  `['no', 'yes', 'unknown']`
- **housing** *(categorical)*: Has housing loan?  
  `['no', 'yes', 'unknown']`
- **loan** *(categorical)*: Has personal loan?  
  `['no', 'yes', 'unknown']`

---

### ☎️ Contact Information (from current campaign)
- **contact** *(categorical)*: Contact communication type  
  `['cellular', 'telephone']`
- **month** *(categorical)*: Last contact month  
  `['jan', 'feb', ..., 'dec']`
- **day_of_week** *(categorical)*: Last contact day  
  `['mon', 'tue', 'wed', 'thu', 'fri']`
- **duration** *(numeric)*: Duration of last contact in seconds  
  ⚠️ *Important:* Duration strongly affects target (`y`), but is **only known after the call**.  
  It should **not be used in predictive models** for real-time classification — include only for benchmarking.

---

### 🔁 Previous Campaign Attributes
- **campaign** *(numeric)*: Number of contacts during current campaign (including last)
- **pdays** *(numeric)*: Days since last contact in previous campaign  
  `999` means "not previously contacted"
- **previous** *(numeric)*: Number of previous contacts
- **poutcome** *(categorical)*: Outcome of the previous campaign  
  `['failure', 'nonexistent', 'success']`

---

### 📉 Economic Context Attributes
- **emp.var.rate** *(numeric)*: Employment variation rate (quarterly)
- **cons.price.idx** *(numeric)*: Consumer price index (monthly)
- **cons.conf.idx** *(numeric)*: Consumer confidence index (monthly)
- **euribor3m** *(numeric)*: Euribor 3-month rate (daily)
- **nr.employed** *(numeric)*: Number of employees (quarterly)

---

## 🎯 Target Variable
- **y** *(binary)*: Has the client subscribed to a term deposit?  
  `['yes', 'no']`


In [1]:
#first import pandas and read csv
import pandas as pd

In [2]:
 # read CSV file
df = pd.read_csv('bank-additional-full.csv', sep=';') 

In [3]:
#lets do an inital check of the data, with head(), info() and describe()
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [4]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

## 📊 Summary Notes: Descriptive Statistics & Observations

---

### 1. `age`
- **Mean**: ~40 years  
- **Range**: 17 to 98  
- **Standard Deviation**: ~10  
- ✔️ Distribution is reasonable for an adult banking population.  
- 📌 Most clients fall within 30–50 years old.

---

### 2. `duration` *(call length in seconds)*
- **Mean**: ~258 seconds (~4 min 18 sec)  
- **Median**: 180 seconds  
- **Range**: 0 to 4918 seconds (~82 min)  
- 🔺 Skewed right: A few long calls pull the average up.  
- ⚠️ *Important:* This value is only known **after** the call → don't use for real-time prediction.

## 📊 Summary Notes: Descriptive Statistics & Observations (Continued)

---

### 3. `campaign` *(number of contacts during this campaign)*  
- **Mean**: ~2.57 contacts  
- **Median**: 2  
- **Max**: 56 (highly skewed)  
- ⚠️ Most clients contacted 1–3 times; a few outliers contacted over 50 times.  
- 🔍 Consider capping or binning to reduce skew impact.

---

### 4. `pdays` *(days since last contact; 999 = never contacted)*  
- **Dominated by 999** values (no prior contact).  
- ✅ Converted to binary feature `prior_contact` (0 = no prior, 1 = prior contact).  
- 🔎 Further analysis on non-999 values possible but optional.

---

### 5. `previous` *(number of contacts before this campaign)*  
- **Mean**: 0.17, mostly zeros.  
- ⚠️ Low variance; may have limited predictive power.  
- Can be combined with `prior_contact` for contact history insights.

---

### 6. `emp.var.rate` *(employment variation rate)*  
- **Range**: -3.4 to 1.4 (economic downturn to growth).  
- ✅ Important macroeconomic indicator likely correlated with subscription rates.

---

### 7. `cons.price.idx` *(consumer price index)*  
- **Narrow range** (92.2 to 94.8), low variance.  
- May influence economic sentiment.

---

### 8. `cons.conf.idx` *(consumer confidence index)*  
- All negative values (-50.8 to -26.9), indicating overall pessimism.  
- ✅ Potentially strong predictor reflecting consumer confidence.

---

### 9. `euribor3m` *(3-month Euribor interest rate)*  
- Range: 0.63 to 5.05; skewed left.  
- ✅ Historically highly predictive for term deposit subscription.

---

### 10. `nr.employed` *(number of employees)*  
- Small range (4963.6 to 5228.1), low variation.  
- Proxy for employment level, useful in conjunction with other economic features.

---

### Overall Summary  
- `duration` is right-skewed and only known post-call, so exclude for real-time prediction.  
- Outliers exist in `campaign` and require handling.  
- `prior_contact` effectively summarizes `pdays`.  
- Economic indicators (`emp.var.rate`, `euribor3m`, `cons.conf.idx`) are key contextual features to leverage.


In [5]:
#Identify outliers
#I'm going to use the IQR method to identify outliers in the dataframe
def identify_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return outliers, lower_bound, upper_bound

# Apply the function to duration, campaign, and pdays
outliers_duration, dur_low, dur_high = identify_outliers_iqr(df, 'duration')   
outliers_campaign, camp_low, camp_high = identify_outliers_iqr(df, 'campaign')
outliers_pdays, pdays_low, pdays_high = identify_outliers_iqr(df, 'pdays')

# Print summary
print(f"Duration outliers: {len(outliers_duration)} (above {dur_high:.2f})")
print(f"Campaign outliers: {len(outliers_campaign)} (above {camp_high:.2f})")
print(f"Pdays outliers: {len(outliers_pdays)} (below {pdays_low:.2f} or above {pdays_high:.2f})")

Duration outliers: 2963 (above 644.50)
Campaign outliers: 2406 (above 6.00)
Pdays outliers: 1515 (below 999.00 or above 999.00)


## 📌 Outlier Summary Table

| Column      | Outlier Count | Issue                              | Recommendation                          |
|-------------|----------------|-------------------------------------|------------------------------------------|
| `duration`  | 2,963          | Right-skewed; long tail of extreme call lengths | 🔹 Cap at 645<br>🔹 Log-transform<br>🔹 Create binary flag `is_long_call` |
| `campaign`  | 2,406          | Clients contacted excessively (e.g. 56 times)   | 🔹 Cap at 6<br>🔹 Create `high_contact` flag<br>🔹 Analyze impact on conversion |
| `pdays`     | 1,515 (≠999)   | 999 is a **placeholder**, not a real value     | 🔹 Replace 999 with `NaN` or similar<br>🔹 Create `was_previously_contacted` flag |
